### PR TITLE
Allow update and create on timestamp fields

### LINKED ISSUE #2135: Allow optional autogeneration of mutations for fields with an @timestamp directive
**Is your feature request related to a problem? Please describe.**
I'm currently dealing with an issue where I have a `DateTime` field where I would like the initial value to be set automatically e.g. ` @timestamp(operations: [CREATE])` but then have the option to manually update the value via the usual auto generated resolvers.

**Describe the solution you'd like**
I think one way of accomplishing this would be to have the signature of the `@timestamp` directive update to look something like the following
```
directive @timestamp(
    """Which events to generate timestamps on. Defaults to both create and update."""
    operations: [TimestampOperation!]! = [CREATE, UPDATE]
    """Whether to auto generate mutation resolvers for the field. Defaults to false"""
    generateResolvers: Boolean
) on FIELD_DEFINITION
```

**Describe alternatives you've considered**
I've looked into the following alternative methods of doing this

- apoc triggers - these aren't supported on Aura
- The `@callback` directive - this isn't supported on `DateTime` fields

### PR DESCRIPTION
**Pull request by mjfwebb**
# Description

A timestamp for creates should be updatable, and a timestamp for updates should be createable. This PR makes that possible.

## Complexity

Medium

Complexity:

Due to the changes required in the AttributeAdapter, there was also a cascade of changes made in the RelationshipAdapter and associated methods.

# Issue

Closes  #2135

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3404 | 3404 | 34 | 43 | 1134 |
|  aggregations.TopLevelAggregateWithMultipleFields | 6802 | 6802 | 43 | 53 | 1134 |
|  aggregations.NestedAggregation | 16544 | 16544 | 71 | 79 | 2174 |
|  aggregations.AggregationWithWhere | 11970 | 11970 | 40 | 67 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116478 | 22116478 | 2116 | 2460 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116478 | 22116478 | 2097 | 2137 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9701 | 9701 | 37 | 44 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9928 | 9928 | 33 | 40 | 2174 |
|  aggregations.DeeplyNestedCount | 11062184 | 11062184 | 2792 | 3030 | 2008534 |
|  aggregations.InterfacesAggregations | 6242 | 6242 | 42 | 54 | 2080 |
|  aggregations.InterfacesAggregationsWithTwoFields | 11444 | 11444 | 69 | 103 | 2080 |
|  batch-create.BatchCreate | 4200 | 4200 | 126 | 169 | 600 |
|  batch-create.BatchCreateSmall | 77 | 77 | 55 | 76 | 11 |
|  connect.createAndConnect | 14424 | 14424 | 157 | 172 | 3003 |
|  connections.Connection | 14049 | 14049 | 55 | 74 | 2174 |
|  connections.NestedConnection | 41068 | 41068 | 109 | 131 | 4516 |
|  create.SimpleMutation | 7 | 7 | 49 | 61 | 1 |
|  cypher-directive.TopLevelMutationDirective | 1135 | 1135 | 28 | 25 | 1134 |
|  delete.SimpleDelete | 19401 | 19401 | 195 | 239 | 1040 |
|  delete.NestedDeleteInUpdate | 19165 | 19165 | 130 | 198 | 1183 |
|  2871.NestedRelationshipFilter | 23117 | 23117 | 52 | 96 | 4441 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 35 | 46 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22891 | 22891 | 74 | 111 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 5201 | 5201 | 33 | 42 | 1040 |
|  2925.NestedSingleRelationshipRequiredFilter | 9361 | 9361 | 50 | 59 | 1040 |
|  query.SimpleQuery | 3121 | 3121 | 19 | 24 | 1040 |
|  query.SimpleQueryWithRelationship | 16129 | 16129 | 37 | 48 | 2174 |
|  query.QueryWhere | 9691 | 9677 | 35 | 43 | 2168 |
|  query.SimpleQueryWithNestedWhere | 9858 | 9844 | 46 | 54 | 2168 |
|  query.Nested | 10095502 | 10095502 | 6794 | 6698 | 2008534 |
|  query.NestedWithFilter | 10074351 | 10074351 | 6646 | 6542 | 2004000 |
|  query.OrFilterOnRelationships | 42616 | 42823 | 148 | 219 | 1696 |
|  query.OrFilterOnRelationshipsAndNested | 35703 | 35679 | 317 | 235 | 1696 |
|  query.QueryWithNestedIn | 13190 | 13729 | 55 | 56 | 1158 |
|  query.NestedConnectionWhere | 9801 | 9801 | 65 | 62 | 2174 |
|  query.DeeplyNestedConnectionWhere | 9838 | 9981 | 100 | 87 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 18776 | 18926 | 218 | 157 | 1592 |
|  query.NestedWithRelationshipSingleFilters | 3865 | 3865 | 154 | 175 | 1134 |
|  query.Fulltext | 64 | 64 | 34 | 43 | 16 |
|  query.FulltextWithNestedQuery | 558 | 558 | 42 | 46 | 84 |
|  sorting-and-cypher.TopLevelSortWithCypher | 14059 | 14059 | 46 | 58 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypher | 14059 | 14059 | 64 | 73 | 2174 |
|  sorting-and-cypher.TopLevelSortWithCypherWithNested | 14200 | 14185 | 62 | 65 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypherWithNested | 14200 | 14185 | 95 | 105 | 2174 |
|  sorting-and-cypher.TopLevelSortWithExpensiveCypher | 14688 | 14591 | 92 | 124 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithExpensiveCypher | 14540 | 14376 | 124 | 117 | 2174 |
|  sorting.SortMultipleTypes | 3498 | 3498 | 81 | 88 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 14478 | 14453 | 101 | 114 | 2174 |
|  sorting.SortOnNestedFields | 14049 | 14049 | 42 | 59 | 2174 |
|  sorting.SortDeeplyNestedFields | 43148 | 43148 | 85 | 106 | 4516 |
|  sorting.ConnectionWithSort | 3281 | 3281 | 63 | 80 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 50 | 69 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 50 | 59 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 243 | 307 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 287 | 301 | 33033 |
|  update.NestedUpdate | 16143 | 16143 | 91 | 109 | 2002 |

</details>
        
Old Schema Generation: 26.806s
Schema Generation: 27.411s
Old Subgraph Schema Generation: 30.267s
Subgraph Schema Generation: 29.916s

File: packages/graphql/tests/schema/issues/2377.test.ts [hunk 1]
```diff
@@ -381,12 +382,14 @@ describe("https://github.com/neo4j/graphql/issues/2377", () => {
             }
 
             input ResourceOnCreateInput {
+              createdAt: DateTime!
```
Comment (darrellwarde) [inline packages/graphql/tests/schema/issues/2377.test.ts:385]: This doesn't seem right, this is essentially equivalent to create input, so you shouldn't be able to set this here.

File: packages/graphql/tests/tck/connections/connect-or-create/connect-or-create.test.ts [hunk 2]
```diff
@@ -178,7 +178,7 @@ describe("Create or Connect", () => {
             typeDefs = gql`
                 type Movie {
                     id: ID! @id @unique
-                    createdAt: DateTime! @timestamp(operations: [CREATE])
+                    createdAt: DateTime @timestamp(operations: [CREATE])
```
Comment (darrellwarde) [inline packages/graphql/tests/tck/connections/connect-or-create/connect-or-create.test.ts:181]: You should be able to leave these as non-nullable!

```suggestion
                    createdAt: DateTime! @timestamp(operations: [CREATE])
```

File: packages/graphql/tests/tck/connections/connect-or-create/connect-or-create.test.ts [hunk 3]
```diff
@@ -463,8 +463,8 @@ describe("Create or Connect", () => {
 
                 interface ActedIn @relationshipProperties {
                     id: ID! @id
-                    createdAt: DateTime! @timestamp(operations: [CREATE])
-                    updatedAt: DateTime! @timestamp(operations: [UPDATE])
+                    createdAt: DateTime @timestamp(operations: [CREATE])
```
Comment (darrellwarde) [inline packages/graphql/tests/tck/connections/connect-or-create/connect-or-create.test.ts:466]: You should be able to leave these as non-nullable!

```suggestion
                    createdAt: DateTime! @timestamp(operations: [CREATE])
```

**Review by darrellwarde — CHANGES_REQUESTED** [review]
The changes look good, but there seems to now be a problem on the create input for `connectOrCreate` operations.

**Pull request merged**